File: ClientApp/src/app/_services/data.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';
import { MedicineDTO } from '../_models/medicineDTO';
import { PatientDTO } from '../_models/patientDTO';
import { DoctorDTO } from '../_models/doctorDTO';

@Injectable({
  providedIn: 'root'
})
export class DataService {

  private url!: string;

  constructor(private httpClient: HttpClient) {
    this.url = `${environment.apiUrl}/data`;
  }

  getMedicines(): Observable<MedicineDTO[]> {
    let fullUrl = this.url + '/medicines';
    return this.httpClient.get<MedicineDTO[]>(fullUrl, { responseType: 'json' });
  }

  getDoctors(): Observable<DoctorDTO[]> {
    let fullUrl = this.url + '/doctors';
    return this.httpClient.get<DoctorDTO[]>(fullUrl, { responseType: 'json' });
  }

  getPatients(): Observable<PatientDTO[]> {
    let fullUrl = this.url + '/patients';
    return this.httpClient.get<PatientDTO[]>(fullUrl, { responseType: 'json' });
  }


}

File: API/Services/DataService.cs
using API.Data;
using API.DTOs;
using API.Interfaces;
using Microsoft.EntityFrameworkCore;

namespace API.Services;

public class DataService : IDataService
{
    private readonly AppDbContext _context;
    public DataService(AppDbContext context)
    {
        _context = context;
    }

    public async Task<List<PatientDTO>> GetPatients()
    {
        var patients = await _context.Patients
            .Select(p => new PatientDTO { Id = p.Id, Name = p.Name })
            .ToListAsync();

        return patients;
    }

    public async Task<List<DoctorDTO>> GetDoctors()
    {
        var doctors = await _context.Doctors
            .Select(d => new DoctorDTO { Id = d.Id, Name = d.Name })
            .ToListAsync();

        return doctors;
    }

    public async Task<List<MedicineDTO>> GetMedicines()
    {
        var medicines = await _context.Medicines
            .Select(m => new MedicineDTO { Id = m.Id, Name = m.Name })
            .ToListAsync();

        return medicines;
    }

}
